addons/sale_stock/models/sale_order.py

```diff
@@ -64,6 +64,12 @@ def _compute_expected_date(self):
                 expected_date = min(dates_list) if order.picking_policy == 'direct' else max(dates_list)
                 order.expected_date = fields.Datetime.to_string(expected_date)
 
+    @api.model
+    def create(self, vals):
+        if 'company_id' in vals and vals.get('company_id') != self.env.company.id:
+            vals['warehouse_id'] = self.env['stock.warehouse'].search([('company_id', '=', vals.get('company_id'))], limit=1).id
```

Comment:
Do we really want to force an arbitrary warehouse in vals without prior notice to the user ? What if `warehouse_id` is already in vals ?
This could probably be done with `warehouse_id` being a stored editable computed field. The user would see the change in the UI